src/include/serial_inline.h

```diff
@@ -79,10 +81,12 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Place a compiler barrier here to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = *ins_stack[i];
+        WT_C11_BARRIER();
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST))
             return (i == 0 ? WT_RESTART : 0);
         if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
             ins_head->tail[i] = new_ins;
```

Comment:
It depends on where the data is defined. If it's on session or cursor, it is thread local. If the data is on btree, it is shared.

Reply:
It is a bit confusing because you can't see what things are on the cursor or session locally in this function. Also, if I understand correctly (not sure that I do!) `ins_stack[i]` itself is stored on the cursor, but it is a pointer to something stored in the btree?

Based on the name, I had assumed that `ins_head` was a pointer to the head of the skiplist which I assumed was stored in the tree. Is that correct? If so, doesn't that mean that this needs to be an atomic access and probably needs a memory ordering?

Reply:
bq. Based on the name, I had assumed that ins_head was a pointer to the head of the skiplist which I assumed was stored in the tree. Is that correct? If so, doesn't that mean that this needs to be an atomic access and probably needs a memory ordering?

In each element in the skip list, there is an array next storing the pointers to the next element at each level in the skip list. ins_head stores the addresses of that array element.

In other words, it stores the pointer to the pointer of a skip list element. The address itself doesn't change even there is concurrent update to the skip list.

Reply:
>  The address itself doesn't change even there is concurrent update to the skip list.

Sure. But I'm specifically referring to the pointer stored at `ins_head->tail[i]`, not just `ins_head` itself. I assume that the `i`th `tail` pointer needs to be accessed with at least store-release and load-acquire semantics.

Reply:
Looks like this was already addressed by https://github.com/wiredtiger/wiredtiger/pull/9010/commits/1ae6a0bc084fa9e65ae8df99605e6b5ea862498a#diff-302b65ca76325e7e93eb73d95d1993931e1b90d3a660d8a5fd4417c7d005ae3eR91-R93